[dist_optim] add distributed functional AdamW optimizer

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50630 [dist_optim] add warning to distributed optimizer
* **#50620 [dist_optim] add distributed functional AdamW optimizer**
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* #50623 [dist_optim] add distributed functional Adadelta optimizer
* #50618 [dist_optim] add distributed functional sgd optimizer
* #50624 [dist_optim] add distributed functional Adam optimizer

Add TorchScript compatible AdamW functional optimizer to distributed optimizer

Differential Revision: [D25932774](https://our.internmc.facebook.com/intern/diff/D25932774)

wanchaol committed:
[dist_optim] add distributed functional AdamW optimizer

Add TorchScript compatible AdamW functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional AdamW optimizer"


Add TorchScript compatible AdamW functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional AdamW optimizer"


Add TorchScript compatible AdamW functional optimizer to distributed optimizer

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional AdamW optimizer"


Add TorchScript compatible AdamW functional optimizer to distributed optimizer

Differential Revision: [D25932774](https://our.internmc.facebook.com/intern/diff/D25932774)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional AdamW optimizer"


Add TorchScript compatible AdamW functional optimizer to distributed optimizer

Differential Revision: [D25932774](https://our.internmc.facebook.com/intern/diff/D25932774)

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_optim] add distributed functional AdamW optimizer"


Add TorchScript compatible AdamW functional optimizer to distributed optimizer

Differential Revision: [D25932774](https://our.internmc.facebook.com/intern/diff/D25932774)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, mainly checked for parity between this and https://github.com/pytorch/pytorch/pull/50411

wanchaol committed:
Update on "[dist_optim] add distributed functional AdamW optimizer"


Add TorchScript compatible AdamW functional optimizer to distributed optimizer

Differential Revision: [D25932774](https://our.internmc.facebook.com/intern/diff/D25932774)

[ghstack-poisoned]
